src/fcmp_pp/tower_cycle.cpp

```diff
@@ -0,0 +1,283 @@
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "string_tools.h"
+#include "tower_cycle.h"
+
+namespace fcmp_pp
+{
+namespace tower_cycle
+{
+//----------------------------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------------------------
+Helios::Point Helios::hash_init_point() const
+{
+    return fcmp_pp_rust::helios_hash_init_point();
+}
+//----------------------------------------------------------------------------------------------------------------------
+Selene::Point Selene::hash_init_point() const
+{
+    return fcmp_pp_rust::selene_hash_init_point();
+}
+//----------------------------------------------------------------------------------------------------------------------
+Helios::CycleScalar Helios::point_to_cycle_scalar(const Helios::Point &point) const
+{
+    return fcmp_pp_rust::helios_point_to_selene_scalar(point);
+}
+//----------------------------------------------------------------------------------------------------------------------
+Selene::CycleScalar Selene::point_to_cycle_scalar(const Selene::Point &point) const
+{
+    return fcmp_pp_rust::selene_point_to_helios_scalar(point);
+}
+//----------------------------------------------------------------------------------------------------------------------
+Helios::Point Helios::hash_grow(
+    const Helios::Point &existing_hash,
+    const std::size_t offset,
+    const Helios::Scalar &existing_child_at_offset,
+    const Helios::Chunk &new_children) const
+{
+    auto result = fcmp_pp_rust::hash_grow_helios(
+        existing_hash,
+        offset,
+        existing_child_at_offset,
+        new_children);
+
+    if (result.err != nullptr)
+    {
+        free(result.err);
+        throw std::runtime_error("failed to hash grow");
+    }
```

Comment:
Rust wont allocate for zero-sized types: https://doc.rust-lang.org/std/boxed/struct.Box.html#method.new

Which means the pointer you are freeing here is invalid and the other places using a ptr to `()`.

Reply:
Shouldn't the pointer be null in that case (making this obviously broken for that reason, as a returned error won't be identified as an error)?

Reply:
https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_raw

Will be non-null. I'm unsure the free behavior here 0_o

Reply:
Added a TODO for this